File: integrators/express/src/cookies.ts
/**
 * Rewrite the `Domain=` attribute of a single `Set-Cookie` header value.
 *
 * - If the cookie contains a `Domain=<value>` attribute (case-insensitive), its
 *   value is replaced with `newDomain`, with any trailing `:port` suffix
 *   stripped first (cookie `Domain` attributes do not include a port).
 * - If `newDomain` is `undefined` (or empty), the entire `; Domain=<value>`
 *   segment is removed so the browser scopes the cookie to the current host.
 * - If no `Domain=` attribute is present, the input is returned unchanged.
 *
 * All other attributes, their casing/order and the original cookie name/value
 * bytes are preserved.
 */
export function rewriteSetCookieDomain(
  setCookie: string,
  newDomain: string | undefined,
): string {
  const domainSegment = /;\s*Domain\s*=\s*[^;]+/i;
  if (!domainSegment.test(setCookie)) return setCookie;

  const stripped = newDomain ? stripPort(newDomain) : '';
  if (!stripped) {
    return setCookie.replace(domainSegment, '');
  }
  return setCookie.replace(
    /(;\s*Domain\s*=\s*)[^;]+/i,
    `$1${stripped}`,
  );
}

/**
 * Apply {@link rewriteSetCookieDomain} to each entry of a `Set-Cookie` array
 * (e.g. the result of `Headers.getSetCookie()`).
 */
export function rewriteSetCookieDomains(
  values: string[],
  newDomain: string | undefined,
): string[] {
  return values.map((value) => rewriteSetCookieDomain(value, newDomain));
}

function stripPort(host: string): string {
  if (host.startsWith('[')) {
    const end = host.indexOf(']');
    return end === -1 ? host : host.slice(1, end);
  }
  const colon = host.indexOf(':');
  return colon === -1 ? host : host.slice(0, colon);
}

File: integrators/fastapi/qelos_integrator_fastapi/cookies.py
from __future__ import annotations

import re

_DOMAIN_SEGMENT = re.compile(r";\s*Domain\s*=\s*[^;]+", re.IGNORECASE)
_DOMAIN_REPLACE = re.compile(r"(;\s*Domain\s*=\s*)[^;]+", re.IGNORECASE)


def strip_port(host: str) -> str:
    if host.startswith("["):
        end = host.find("]")
        return host if end == -1 else host[1:end]
    colon = host.find(":")
    return host if colon == -1 else host[:colon]


def rewrite_set_cookie_domain(set_cookie: str, new_domain: str | None) -> str:
    """Rewrite the ``Domain=`` attribute of a single ``Set-Cookie`` header value.

    Semantics match ``integrators/nuxt/src/cookies.ts``.
    """
    if not _DOMAIN_SEGMENT.search(set_cookie):
        return set_cookie

    if new_domain is None or not str(new_domain).strip():
        return _DOMAIN_SEGMENT.sub("", set_cookie)

    stripped = strip_port(str(new_domain).strip())
    if not stripped:
        return _DOMAIN_SEGMENT.sub("", set_cookie)
    return _DOMAIN_REPLACE.sub(r"\1" + stripped, set_cookie, count=1)


def rewrite_set_cookie_domains(values: list[str], new_domain: str | None) -> list[str]:
    """Apply :func:`rewrite_set_cookie_domain` to each ``Set-Cookie`` value."""
    return [rewrite_set_cookie_domain(v, new_domain) for v in values]
